### ISSUE TITLE
Header forwarding for stock networking redirects

### ISSUE THREAD

**Issue by philipgiuliani** [bumptech/glide#536]
Im loading an image into my imageView by using GlideUrl. To get the image i need to send an authorization header to the server. The Server responds correctly and redirects the link to a s3 url.

**Works correctly:**

``` java
Glide.with(this)
    .load("serverurl.jpg")
    .into(imageView);
```

**Does not follow the redirect:**

``` java
GlideUrl url = new GlideUrl("serverurl.jpg", new LazyHeaders.Builder()
    .setHeader("Authorization", "Token token=myAuthToken")
    .build());

Glide.with(this)
    .load(url)
    .into(imageView);
```

**Comment by TWiStErRob** [comment]
Are you using an integration lib (okhttp/volley)? Also which Glide version?

**Comment by philipgiuliani** [comment]
Sorry for the missing details.

Glide Version: 3.6.1

No i dont have OkHttp/Volley as dependencies.
But i have Retrofit and Picasso there which also depend on OkHttp.

**Comment by TWiStErRob** [comment]
That's OK to have them on the classpath; as long as you don't have the [glue](https://github.com/bumptech/glide/wiki/Integration-Libraries) to connect it to Glide, you're using the default Android networking.

No worries about the missing info, one more thing I forgot to ask: phone and Android version you're running on it?
(Actually all these questions are listed at the [issue template](https://github.com/bumptech/glide/blob/master/CONTRIBUTING.md), I should've asked for that :)

**Comment by philipgiuliani** [comment]
Device: Samsung Galaxy S5
Android: 5.1
OS: Cyanogenmod 12

Is there any simple way to debug the Glide requests? Haven't found anything!

**Comment by TWiStErRob** [comment]
See https://github.com/bumptech/glide/wiki/Debugging-and-Error-Handling, thought it's a little scarce.

I would go like this to figure out what's wrong:
1. Turn on `GenericRequest` logging. If that doesn't print anything useful:
2. Add a `.listener()` to both requests and Log all arguments. If there's no difference for both loads:
3. Put a breakpoint in `HttpUrlFetcher.loadDataWithRedirects` and see what's different (variables/execution path) with the two loads.

**Comment by philipgiuliani** [comment]
Hi i found the problem i guess. My Server is redirecting with "302 Found" to an Amazon S3 URL. The problem is that the header "Authorization Token token=mytoken" is passed to the Amazon S3 and that returns:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<Error>
  <Code>InvalidArgument</Code>
  <Message>Unsupported Authorization Type</Message>
  <ArgumentName>Authorization</ArgumentName>
  <ArgumentValue>Token token=mytoken</ArgumentValue>
  <RequestId>108E0E112DAE402C</RequestId>
  <HostId>t17TFtdFQYacJAH1LdMMWyD9h3x7gmkPmmsvs19Ghl0/Dt1aulblx+cIY5mibKmHjZfvLwnhFbQ=</HostId>
</Error>
```

You have any idea how to dont pass the header on after a redirect? Is that already possible in some way?

Thanks.

**Comment by TWiStErRob** [comment]
Nice find. Let's see what @sjudd says, this may be considered a bug.

In the meantime you have two options:
* hide the redirect inside your server, that is return the stream not a 302. This also saves a roundtrip for each image... think 3G users.
* Replace the default HttpUrlFetcher with your own copy which only applies headers when lastUrl is null. This replacement should go down the same way you would write an integration library, so look at how [OkHttp is done](https://github.com/bumptech/glide/tree/3.0/integration/okhttp/src/main/java/com/bumptech/glide/integration/okhttp) (don't forget to check AndroidManifest.xml for GlideModule registration)

**Comment by sjudd** [comment]
Is there any convention here? Do you always only add auth headers before the first redirect, or is this specific to S3?

**Comment by philipgiuliani** [comment]
No this is just for my server. The server has to check if this user is permitted to see this image, and if he his, he redirects it to an S3 url.

**Comment by sjudd** [comment]
Ah ok, then as @TWiStErRob mentioned your best bet is to copy/modify the existing HttpUrlFetcher, or one of the integration libraries.

**Comment by TWiStErRob** [comment]
@sjudd I think it's reasonable to believe that the convention is that only the first request needs the headers, or at least while we're on the same domain. The problem is that this behavior of the headers may be different for different header keys. For example you don't want to lose the User-Agent on any of them requests, but an Accept may be needed, may be not the second time.
Can you think of a way to nicely generalize and allow to set headers for specific redirects? Though that may be an overkill.

Regardless, a flag to send the same headers all the time or just one time (`n` times?) may be nice mid-way solution and I think would help this specific issue unless @philipgiuliani needs some headers for S3 too.

**Comment by philipgiuliani** [comment]
I did the changes on the server currently to stream the images from S3 to the client. In our case not perfect because we switched to S3 to reduce server load (the application is completely image based).
I do not need to pass headers to S3, just to my server. I think the best way is to pass the headers just on the first request. I dont think they are still needed in a second request normally. Flag for that would be the best. Great work and thanks.

**Comment by TWiStErRob** [comment]
@sjudd Hmm, actually it would be interesting to know how this works with OkHttp and Volley, how they handle redirected headers.

@philipgiuliani and that's another possible workaround for   you, to add an integration lib and restore your server to 302, maybe these libs handle it the way you need it. You said you have OkHttp already on classpath, so you just need the 3 Glide glue classes.

**Comment by sjudd** [comment]
I'm not able to find any standard way of including headers from re-directs. If we get a few upvotes on this or other people file similar issues we might be able to add a mechanism to GlideUrl that would allow forwarding of arbitrary headers on 302, but I'm not sure it's worth it at this point.

I'd also be curious to know if anyone experiencing this has gotten it to work with volley or okhttp.

**Comment by Nursik91** [comment]
Hi guys! I faced the same problem. I tried to load from our server which redirects me to s3. And glide cannot load it. I think I have the same case. Did you solve it?

**Comment by mirhec** [comment]
I'm having the same issue right now. @nursike did you find a solution?

**Comment by mirhec** [comment]
If someone has the same problem, here's my workaround:

I changed my Api to accept the Auth-Token via query string, so I can call it like `/your/endpoint?token=your-token`. But this would lead to the problem that the file wouldn't be cached properly, because the token changes quite often. So I used to use the class [`GlideUrlWithQueryParameter`](https://futurestud.io/tutorials/glide-how-to-ignore-url-query-parameters-in-cache-lookup) to exclude this query param from being used to dectect url changes. This works quite well.
